File: java/app/src/main/java/com/leecampbell/cod/application/FakeRepository.java
package com.leecampbell.cod.application;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import com.leecampbell.cod.domain.contracts.*;
import com.leecampbell.cod.domain.model.Loan;
import com.leecampbell.cod.domain.services.*;

final class FakeRepository implements Repository {
    private final ArrayList<DomainEvent> committedEvents = new ArrayList<>();

    public Loan get(UUID id) {
        return new Loan(id);
    }

    public void save(Loan item) {
        List<DomainEvent> temp = Arrays.asList(item.getUncommittedEvents());
        committedEvents.addAll(temp);

        item.clearUncommittedEvents();
    }
}

File: typescript/packages/domain-tests/src/fake-repository.ts
import {
  AggregateRoot,
  Event,
  type Repository,
} from "@cod/domain";

export class FakeRepository<T extends AggregateRoot>
  implements Repository<T>
{
  readonly #items = new Map<string, T>();
  readonly #factory: (id: string) => T;
  readonly committedEvents: Event[] = [];

  constructor(factory: (id: string) => T) {
    this.#factory = factory;
  }

  async get(id: string): Promise<T> {
    let item = this.#items.get(id);
    if (!item) {
      item = this.#factory(id);
      this.#items.set(id, item);
    }
    return item;
  }

  async save(item: T): Promise<void> {
    this.committedEvents.push(...item.getUncommittedEvents());
    item.clearUncommittedEvents();
  }
}
